### PR TITLE
add github workflows for DR test

### PR DESCRIPTION
**Pull request by RMcVelia**
## Context

Create and/or update workflows so we can use workflows for DR and DR testing

## Changes proposed in this pull request

- update the database-workflow to use the backup-postgres github action
- add postgres-ptr workflow that uses the ptr-postgres github action
- add postgres-restore workflow that uses the restore-postgres-backup github action

Enable above workflows to run on demand and add the staging env as an option for the DR test

## Guidance to review

staging backup: https://github.com/DFE-Digital/itt-mentor-services/actions/runs/10487024842/job/29046518464 

## Link to Trello card

https://trello.com/c/Yy7qyj4N/1985-ittms-create-dr-test-github-workflows

### CODE REVIEW

**Comment by RMcVelia** [comment]
[ITTMS create DR test github workflows](https://trello.com/c/Yy7qyj4N/1985-ittms-create-dr-test-github-workflows)

File: .github/workflows/database-backup.yml [hunk 1]
```diff
@@ -4,72 +4,69 @@ concurrency: build_and_deploy_main
...
+        echo "DB=${AZURE_RESOURCE_PREFIX}-${SERVICE_SHORT}-${CONFIG_SHORT}-pg" >> $GITHUB_ENV
 
-    - name: Set up environment variables
-      shell: bash
+    - name: Set PTR variables
+      if: ${{ github.event.inputs.PTRdatabase == 'true' }}
       run: |
-        echo "STORAGE_ACCOUNT_RG=s189p01-ittms-pd-rg" >> $GITHUB_ENV
-        echo "STORAGE_ACCOUNT_NAME=s189p01ittmsdbbkppdsa" >> $GITHUB_ENV
+        echo "PTR_DB=${{ env.DB }}-PTR" >> $GITHUB_ENV
```
Comment (saliceti) [inline .github/workflows/database-backup.yml:60]: Can we rename to PTR_DB_SERVER for clarity?

File: .github/workflows/database-backup.yml [hunk 2]
```diff
@@ -4,72 +4,69 @@ concurrency: build_and_deploy_main
...
-        echo "AZURE_STORAGE_CONNECTION_STRING=$STORAGE_CONN_STR" >> $GITHUB_ENV
-
-    - name: Upload Backup to Azure Storage
-      run: |
-        az config set extension.use_dynamic_install=yes_without_prompt
-        az config set core.only_show_errors=true
-        az storage azcopy blob upload --container database-backup \
-        --source ${BACKUP_FILE_NAME}.tar.gz
+    - name: Backup ${{ env.DEPLOY_ENV }} postgres
+      if: ${{ github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && github.event.inputs.overwriteThisMorningsBackup == 'true' ) }}
```
Comment (saliceti) [inline .github/workflows/database-backup.yml:63]: Is this required at the step level? It's already set at job level

File: .github/workflows/postgres-ptr.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      confirm-production:
+        description: Set to true to restore nightly backup if production
+        required: true
+        default: 'false'
+        type: choice
+        options:
+        - 'false'
+        - 'true'
+      restore-time:
+        description: Restore point in time. e.g. 2024-07-24T06:00:00
```
Comment (saliceti) [inline .github/workflows/postgres-ptr.yml:24]: Is it UTC? or GMT? or... ?

File: .github/workflows/postgres-ptr.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    - name: Set environment variables
+      run: |
+        source global_config/${{ inputs.environment }}.sh
+        tf_vars_file=${{ env.TF_VARS_PATH }}/${{ inputs.environment }}.tfvars.json
+        echo "CLUSTER=$(jq -r '.cluster' ${tf_vars_file})" >> $GITHUB_ENV
+        echo "RESOURCE_GROUP_NAME=${AZURE_RESOURCE_PREFIX}-${SERVICE_SHORT}-${CONFIG_SHORT}-rg" >> $GITHUB_ENV
+        echo "DB=${AZURE_RESOURCE_PREFIX}-${SERVICE_SHORT}-${CONFIG_SHORT}-pg" >> $GITHUB_ENV
+
+    - name: Restore ${{ inputs.environment }} postgres
+      if: ${{ inputs.environment != 'production' || (inputs.environment == 'production' && github.event.inputs.confirm-production == 'true' )  }}
```
Comment (saliceti) [inline .github/workflows/postgres-ptr.yml:51]: Can this be moved to job level?

File: .github/workflows/postgres-ptr.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      environment:
+        description: Environment to restore
+        required: true
+        default: staging
+        type: choice
+        options:
+        - 'staging'
+#        - 'production'
+      confirm-production:
+        description: Set to true to restore nightly backup if production
```
Comment (saliceti) [inline .github/workflows/postgres-ptr.yml:16]: Not nightly

File: .github/workflows/postgres-restore.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        echo "BACKUP_FILE=${SERVICE_SHORT}_${CONFIG_SHORT}_${TODAY}.sql" >> $GITHUB_ENV
+        if [ "${{ inputs.backup-file }}" != "" ]; then
+          BACKUP_FILE=${{ inputs.backup-file }}
+        else
+          BACKUP_FILE=${SERVICE_SHORT}_${CONFIG_SHORT}_${TODAY}.sql.gz
+        fi
+        echo "BACKUP_FILE=$BACKUP_FILE" >> $GITHUB_ENV
+
+    - name: Restore ${{ inputs.environment }} postgres
+      if: ${{ inputs.environment != 'production' || (inputs.environment == 'production' && github.event.inputs.confirm-production == 'true' )  }}
```
Comment (saliceti) [inline .github/workflows/postgres-restore.yml:60]: Can this be moved to job level

File: .github/workflows/postgres-ptr.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
+name: Restore Database from point in time
```
Comment (saliceti) [inline .github/workflows/postgres-ptr.yml:1]: This looks like the restore workflow but it's very different. Can the name be clearer, like "Restore database from point in time to new database server"

File: .github/workflows/database-backup.yml [hunk 8]
```diff
@@ -4,72 +4,69 @@ concurrency: build_and_deploy_main
...
+
 jobs:
   backup:
-    name: Backup AKS Database (production)
+    name: Backup AKS Database
     if: ${{ github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && github.event.inputs.overwriteThisMorningsBackup == 'true') }}
     runs-on: ubuntu-latest
     environment:
-        name: production
+      name: ${{ inputs.environment || 'production' }}
```
[the remark this answers is not in the document]
Comment (RMcVelia) [reply .github/workflows/database-backup.yml:38]: inputs only work for workflow_dispatch. Need to add a default for the scheduled run.

File: .github/workflows/database-backup.yml (hunk 2 above)
Comment (RMcVelia) [reply .github/workflows/database-backup.yml]: removed from lower level

File: .github/workflows/postgres-ptr.yml (hunk 3 above)
Comment (RMcVelia) [reply .github/workflows/postgres-ptr.yml]: added UTC

File: .github/workflows/postgres-ptr.yml (hunk 4 above)
Comment (RMcVelia) [reply .github/workflows/postgres-ptr.yml]: moved

File: .github/workflows/postgres-ptr.yml (hunk 5 above)
Comment (RMcVelia) [reply .github/workflows/postgres-ptr.yml]: updated description

File: .github/workflows/postgres-restore.yml (hunk 6 above)
Comment (RMcVelia) [reply .github/workflows/postgres-restore.yml]: moved

File: .github/workflows/postgres-ptr.yml (hunk 7 above)
Comment (RMcVelia) [reply .github/workflows/postgres-ptr.yml]: changed name

File: .github/workflows/database-backup.yml [hunk 9]
```diff
@@ -4,72 +4,68 @@ concurrency: build_and_deploy_main
...
-        bin/konduit.sh -t 7200 itt-mentor-services-production -- pg_dump -E utf8 --clean --if-exists --no-owner --verbose --no-password -f ${BACKUP_FILE_NAME}.sql
-        tar -cvzf ${BACKUP_FILE_NAME}.tar.gz ${BACKUP_FILE_NAME}.sql
+        source global_config/${{ env.DEPLOY_ENV }}.sh
+        tf_vars_file=${{ env.TF_VARS_PATH }}/${{ env.DEPLOY_ENV }}.tfvars.json
+        echo "CLUSTER=$(jq -r '.cluster' ${tf_vars_file})" >> $GITHUB_ENV
+        echo "RESOURCE_GROUP_NAME=${AZURE_RESOURCE_PREFIX}-${SERVICE_SHORT}-${CONFIG_SHORT}-rg" >> $GITHUB_ENV
+        echo "STORAGE_ACCOUNT_NAME=${AZURE_RESOURCE_PREFIX}${SERVICE_SHORT}dbbkp${CONFIG_SHORT}sa" >> $GITHUB_ENV
+        TODAY=$(date +"%F")
+        echo "BACKUP_FILE=${SERVICE_SHORT}_${CONFIG_SHORT}_${TODAY}.sql" >> $GITHUB_ENV
+        echo "DB=${AZURE_RESOURCE_PREFIX}-${SERVICE_SHORT}-${CONFIG_SHORT}-pg" >> $GITHUB_ENV
```
Comment (saliceti) [inline .github/workflows/database-backup.yml:55]: Could we rename to DB_SERVER for clarity and consistency?

Comment (RMcVelia) [reply .github/workflows/database-backup.yml]: done

**Review by saliceti — APPROVED** [review]

**Pull request merged**